keps/77-dynamically-sized-jobs/README.md

```diff
@@ -0,0 +1,260 @@
+- Workload Slices will be created by Kueue and use identical PodTemplates which is already enforced by Kuberay
+- Workload Slices will beneed to belong to the same resource flavor as the top-level RayCluster that was initially admitted
+
+
+### Creating Workload Slices
+
+The [GenericJob interface](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go#L30-L55) will be updated to handle resize operations of jobs.
+
+```golang
+type GenericJob interface {
+  ...
+  ...
+  ResizeJob(wl *kueue.Workload) error
+}
+```
+Jobs implementing the ResizeJob method will create a Workload Slice for every new replica of a job. 
+
+On scale down to M we will change the original Workload's resources and then on scale up to N we will create N-M WorkloadSlice objects that go through admission and scheduling gate checks, i.e. the Workload can "lose" the original (first time admission) quota when scaling down before scaling up.
+
+### Pod Scheduling Gates
+
+Inside **raycluster_webhook** implement schedulingGate injection for pods on RayCluster creation time.
+The Pods will be ungated following a similar behavior as to how a job is suspended and then unsuspended in the when admitted.
+When the RayCluster scales up, the new pods will be gated due to the schedulingGates injection in the webhook.
+
+After the creation of each individual Workload Slice and admission of a Workload Slice, the **workload_scheduling_gates_controller** should be in charge of removing the scheduling gates from each pod. All worker pods from the same worker group share the same pod template, so we only need to ungate the number of pods to match the number of admitted pods, this should be a counter. We don’t want to accidentally ungate too many pods since race conditions could happen and we also don’t want to double count. It's worth mentioning that for the case of recreated pods (i.e. machine failure for example), these pods will go through the admission/scheduuling check again, Kueue is responsible fo removing the scheduling gates when there's available quota and resources to spend on the RayCluster. 
+
+### Garbage Collecting Workload Slices
+
+The logic of folding and deleting would be isolated in this controller. We don’t necessarily have to say that this Workload Slice belongs to a specific pod. This controller will look at the Workload objects and check whether they have the Admitted condition or not.
+
+1. The controller increments the `.status.admission.podSetAssignments.count` and passes the UID of the workload you are folding to the parent workload.
+If we still don’t see the workload being deleted, we at least know it has been counted towards the parent workload and the controller shouldn't fold it again.
+
+## Phases for MVP (alpha)
+
+### Phase 1 - Scale Down
+Scaling down will be the first phase towards MVP because it can be implemented without introducing the Workload Slices. 
```

Comment:
Mentioning the differences between the Dynamically Scaling down (this feature) and Dynamically Resource Reclaiming.
As my understanding, the Dynamically Scaling down doesn't release the deleted pod's resource quota, right?

I understand that we currently don't support Dynamically Resource Reclaiming in the RayCluster, but I think that clarifying the differences would be valuable in this phase.

Reply:
dynamic reclaim, for context https://github.com/kubernetes-sigs/kueue/tree/main/keps/78-dynamically-reclaiming-resources

In theory both can work together.

> As my understanding, the Dynamically Scaling down doesn't release the deleted pod's resource quota, right?

It does.

Reply:
> As my understanding, the Dynamically Scaling down doesn't release the deleted pod's resource quota, right?

@tenzen-y Here's a discussion about that https://github.com/kubernetes-sigs/kueue/pull/1851#discussion_r1537672669

Reply:
> > As my understanding, the Dynamically Scaling down doesn't release the deleted pod's resource quota, right?
> 
> 
> @tenzen-y Here's a discussion about that https://github.com/kubernetes-sigs/kueue/pull/1851#discussion_r1537672669

It seems that the pointing discussion is slightly different.
I meant that can we enable both this feature (KEP-77) and dynamically reclaiming resources (KEP-78).

Reply:
Yes, you could enable both.